MB-4141 Check orders and loa presence

gidjin opened the pull request:
## Description

The story MB-4141 is about handling various error states for payment request creation. This PR covers MB-5345 when the Orders on the MTO is incomplete. Currently it checks for `OriginDutyStation` and `TAC`. This may need to be expanded later if we need to look at other fields, like `NewDutyStation` or `SAC`. This PR didn't check `NewDutyStation` as it is currently required by the data model and database. `SAC` is not used by the payment request process so it's also not checked.  If there is an invalid order, this will return a 409 Conflict because Orders is not an input from the Prime. See https://github.com/transcom/mymove/wiki/API-Errors#http-status-codes

## Reviewer Notes

What will the prime due if they get one of these? Contact the TOO? Where would this be documented? Is BadDataError right here or should it be InvalidCreateInputError so we return a 422 instead.

## Setup

Setup server

```sh
make db_dev_e2e_populate
make server_run
make client_run
```

* Once the server is up, create a new move via the ui and submit it.
* Log in as the too
* Approve the shipment and services, but do not update the orders

_Not sure how to test not having OriginDutyStation without changing the data by hand_

Save this json to a payload.json but update the ids to match the move you created above
```json
{
  "body": {
    "isFinal": false,
    "moveTaskOrderID": "9c7b255c-2981-4bf8-839f-61c7458e2b4d",
    "serviceItems": [
      {
        "id": "94bc8b44-fefe-469f-83a0-39b1e31116fb"
      },
      {
        "id": "eee4b555-2475-4e67-a5b8-102f28d950f8"
      },
      {
        "id": "6431e3e2-4ee4-41b5-b226-393f9133eb6c"
      },
      {
        "id": "fd6741a5-a92c-44d5-8303-1d7f5e60afbf"
      }
    ]
  }
}
```

Try to create a new payment request see the error
```sh
prime-api-client --insecure create-payment-request --filename payload.json | jq
{
  "Payload": {
    "detail": "id: 24a09d26-1650-4ed4-8ce6-7ada81bafc8e is in a conflicting state Orders on MoveTaskOrder (ID: 171ca931-8ad8-43a1-8cf9-29eb2be2fd5c) missing Lines of Accounting TAC",
    "instance": "9295590b-6ac3-4e4d-81bd-feae3c733c19",
    "title": "Conflict Error"
  }
}

```

Validate handler returns correct error a `409 Conflict`:

```sh
❯ prime-api-client --insecure create-payment-request --filename tmp/payloads/local_create_payment_request.json -v
POST /prime/v1/payment-requests HTTP/1.1
Host: primelocal:9443
User-Agent: Go-http-client/1.1
Content-Length: 152
Accept: application/json
Content-Type: application/json
Accept-Encoding: gzip

{"isFinal":false,"moveTaskOrderID":"171ca931-8ad8-43a1-8cf9-29eb2be2fd5c","serviceItems":[{"id":"9b71a6de-cfc1-47da-ad49-30badb7f6bed","params":null}]}

HTTP/1.1 409 Conflict
Content-Length: 254
Cache-Control: no-cache, no-store, must-revalidate
Content-Security-Policy: frame-ancestors 'none'
Content-Type: application/json
Date: Wed, 11 Nov 2020 20:55:02 GMT
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Frame-Options: deny
X-Xss-Protection: 1; mode=block

{"detail":"id: 24a09d26-1650-4ed4-8ce6-7ada81bafc8e is in a conflicting state Orders on MoveTaskOrder (ID: 171ca931-8ad8-43a1-8cf9-29eb2be2fd5c) missing Lines of Accounting TAC","instance":"46e6cb05-5d10-4557-8112-a4090aeeab39","title":"Conflict Error"}

{"Payload":{"detail":"id: 24a09d26-1650-4ed4-8ce6-7ada81bafc8e is in a conflicting state Orders on MoveTaskOrder (ID: 171ca931-8ad8-43a1-8cf9-29eb2be2fd5c) missing Lines of Accounting TAC","instance":"46e6cb05-5d10-4557-8112-a4090aeeab39","title":"Conflict Error"}}% 

```

## Code Review Verification Steps

* [X] Request review from a member of a different team.
* [X] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-4141) for this change
* [Jira subtask](https://dp3.atlassian.net/browse/MB-5345) for this change
* [API error code docs](https://github.com/transcom/mymove/wiki/API-Errors#http-status-codes)

robot-mymove commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:link: <a href="https://dp3.atlassian.net/browse/MB-4141">MB-4141</a></td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 28107410324aec1a0c153f684db6022fbfc17a73
</p>

shimonacarvalho reviewed (approved):
Approving the error, don't have enough context for the LOA change though.

pearl-truss commented on pkg/services/payment_request/payment_request_creator.go:
```diff
@@ -218,6 +221,17 @@ func (p *paymentRequestCreator) createPaymentRequestSaveToDB(tx *pop.Connection,
+	// Verify the Orders on the MTO
+	tx.Load(&moveTaskOrder, "Orders")
+	// Verify that the Orders has LOA
+	if moveTaskOrder.Orders.TAC == nil || *moveTaskOrder.Orders.TAC == "" {
```
For the LOA segments, the EDI generator only does a nil check. Does this matter or should it also check for empty strings?

gidjin replied:
According to the model empty string is a possible choice, we may want to have the EDI generator also check, but I'm not sure how in the ui the TOO would set blank because TAC is a required field so may be moot.

gidjin replied:
It was easy to do so I updated the generator to look for a blank TAC as well as a nil TAC.

reggieriser commented on pkg/services/invoice/ghc_payment_request_invoice_generator_test.go:
```diff
@@ -563,6 +564,29 @@ func (suite *GHCInvoiceSuite) TestNilValues() {
+		_, err := generator.Generate(nilPaymentRequest, false)
+		suite.Error(err)
+		_, ok := err.(services.ConflictError)
+		suite.Equal(true, ok)
```
Just an FYI ... I believe there's a `suite.IsType` that I think does the same thing as the two lines above.

reggieriser commented on pkg/services/payment_request/payment_request_creator_test.go:
```diff
@@ -323,6 +323,97 @@ func (suite *PaymentRequestServiceSuite) TestCreatePaymentRequest() {
+	})
+
+	suite.T().Run("Given move with orders blank LOA, the create should fail", func(t *testing.T) {
+		mtoInvalidOrders := testdatagen.MakeMove(suite.DB(), testdatagen.Assertions{})
```
Nit: This test and the previous tests may be able to be combined/simplified into a table-driven test since it seems we're just checking two different values for the same scenario.

reggieriser reviewed (approved):
I added a few nit comments, but this LGTM.  Going to go ahead and approve as I don't see the things I mentioned as blockers. 👍 

I ran the scenario and got the Conflict Error as expected.  I did notice in the testing notes you said to look for a `400 Bad Request`, but I got a `409 Conflict` like is shown in the response.  I assume the 400 is just a typo in the instructions.

gidjin commented:
> I added a few nit comments, but this LGTM. Going to go ahead and approve as I don't see the things I mentioned as blockers. 👍
> 
> I ran the scenario and got the Conflict Error as expected. I did notice in the testing notes you said to look for a `400 Bad Request`, but I got a `409 Conflict` like is shown in the response. I assume the 400 is just a typo in the instructions.

Nice catch, it was originally a 400 error, I missed that one when I updated the PR.

As for the test feedback, I'll apply to the other PR #5206 since that is dependent on these changes anyway.
